### PR TITLE
webapp: consolidate exception treatment in HTTP handlers

### PR DESCRIPTION
**Pull request by jgehrcke**
For issue #1181.
Ref docs: https://flask.palletsprojects.com/en/2.2.x/errorhandling/#generic-exception-handlers

### CODE REVIEW

File: conbench/api/index.py [hunk 1]
```diff
@@ -207,5 +207,21 @@ def wipe_db():
...
+        f.abort(400, description="foo")
+
+    @api.route("/raise-unexpected", methods=("GET",))
+    def raise_unexpected():
+        """
+        For local development / developer productivity.
+        """
+        raise Exception("doh")
+        # raise AssertionError("doh!")
+        # assert True == False
```
Comment (jonkeane) [inline conbench/api/index.py]: Can we remove these?

**Review by jonkeane — APPROVED** [review]
Thanks for this!

File: conbench/__init__.py [hunk 2]
```diff
@@ -166,32 +169,48 @@ def _init_api_docs(application):
...
-    # part of requently called error handler (even if the import machinery is
-    # fast and cached) is a code smell -- let's see about this.
     import flask as f
 
-    data = {"code": exc.code, "name": exc.name}
+    # For now: even in 'prod', expose the traceback corresponding to the
+    # exception in an HTTP 500 response, facilitating debugging; so that users
+    # can send tracebacks right away.
+    tbstr = "".join(traceback.format_exception(exc))
+    return f.make_response((f"unexpected exception, please report this:\n{tbstr}", 500))
```
Comment (austin3dickey) [inline conbench/__init__.py:184]: "unexpected exception" is very fun to read and say out loud

File: conbench/tests/app/_asserts.py [hunk 3]
```diff
@@ -52,7 +52,8 @@ def assert_404_not_found(self, r):
         # TODO: How to make this an HTML error page?
         # While still having json error rs for the API.
         assert r.content_type == "application/json", r.content_type
-        assert r.json == {"code": 404, "name": "Not Found"}, r.json
+        assert r.json["code"] == 404, "name"
```
Comment (austin3dickey) [inline conbench/tests/app/_asserts.py]: I think "name" isn't intentional here, right?

**Review by austin3dickey — APPROVED** [review]
very nice!

File: conbench/__init__.py (hunk 2 above)
Comment (jgehrcke) [reply conbench/__init__.py:184]: > "unexpected exception" is very fun to read and say out loud

Yes and one of the most powerful concepts to embrace is to -- from the get-go when starting into an engineering problem -- distinguish "expected errors" from "unexpected errors" or to at least think in those two categories in _every part of the code_ (regardless of how one wants to call those).

Especially in distributed systems there's a pretty large set of expected errors along the way, but usually always when there is a human/machine boundary (user data validation), etc. As simple as powerful. I've heard people use "error" vs. "failure" but I find that silly :D The important thing is to actively distinguish.

File: conbench/tests/app/_asserts.py (hunk 3 above)
Comment (jgehrcke) [reply conbench/tests/app/_asserts.py]: Haha, no. I wonder how it got there, and what the facial expression would have been to discover this later.

File: conbench/api/index.py (hunk 1 above)
Comment (jgehrcke) [reply conbench/api/index.py]: Yes! (forgot to push that commit before :shrimp: (and the `:shrimp:` is what github sometimes suggests when I want to write `:shrug:`))

File: conbench/tests/api/_asserts.py [hunk 4]
```diff
@@ -60,7 +60,8 @@ def assert_400_bad_request(self, r, message):
     def assert_401_unauthorized(self, r):
         assert r.status_code == 401, r.status_code
         assert r.content_type == "application/json", r.content_type
-        assert r.json == {"code": 401, "name": "Unauthorized"}, r.json
+        assert r.json["code"] == 401, r.json
+        assert r.json["name"] == "Unauthorized"
```
Comment (jgehrcke) [inline conbench/tests/api/_asserts.py:64]: Note: now the dict also has a `description` key and it's a long  sentence with a description of what this error means; default Flask behavior. Didn't want to put this sentence into the test suite.

File: conbench/api/index.py (hunk 1 above)
Comment (jonkeane) [reply conbench/api/index.py]: Yum 🦐 , now I'm hungry 😂

**Pull request merged**